File: summer-school/MPI/Exercise_05/Skeleton_05.f90
program ghost_cells
    use mpi
    implicit none

    integer, parameter :: N=20
    integer :: my_rank, nprocs, ierr
    integer :: i,j
    integer :: rem, num_local_col
    integer :: proc_right, proc_left
    integer, allocatable :: matrix(:,:)
    integer status1(MPI_Status_size), status2(MPI_Status_size)

    ! Initialize the environment and store the size in "nprocs" and the rank in "my_rank"
    ...

    !  number of columns for each mpi task. How does it work? Try to write it down.
    rem= mod(N,nprocs)
    num_local_col = (N - rem)/nprocs

    if(my_rank < rem) num_local_col = num_local_col+1

    ! Allocation of the global matrix . Instead of "..." put one of the following:
    1. num_local_col  2. num_local_col+1  3. num_local_col+2

    allocate(matrix(N,...))

    ! inizialization of the local matrix
    matrix = my_rank

    ! Information about the neighbour processes (imagine that you are sending left and rught arrays of data, so the variables are named to reflect that)
    proc_right = my_rank+1
    proc_left = my_rank-1
    if(proc_right .eq. nprocs) proc_right = 0
    if(proc_left < 0) proc_left = nprocs-1

    ! check printings
    write(*,*) "my_rank, proc right, proc left ", my_id, proc_right, proc_left
    write(*,*) "my_rank, num_local_col ", my_rank, num_local_col
    write(*,*) "my_rank, matrix(1,1), matrix(1,num_local_col+2), matrix(N,num_local_col+2)", &
                my_rank, matrix(1,1), matrix(1,num_local_col+2), &
                matrix(N,num_local_col+2)

    ! send receive of the ghost regions
    ! First one: sends the leftiest real column to the proc on its left, and receives on the rightest ghost column from the proc on its right
    call mpi_sendrecv(...)

    ! Second one: sends the rightiest real column to the proc on its right, and receives on the leftiest ghost column from the proc on its left
    call mpi_sendrecv(...)

    ! check printings
    write(*,*) "my_rank ", my_rank, " colonna arrivata da sinistra: ", matrix(:,1)
    write(*,*) "my_rank ", my_rank, " colonna arrivata da destra: ", &
                matrix(:,num_local_col+2)

    deallocate(matrix)
    ! Finalize MPI environment
    call mpi_finalize(...)

end program
  

File: summer-school/MPI/Exercise_05/Skeleton_05.c
#include <stdlib.h>
  #include <stdio.h>
  #include <mpi.h>

  #define N 20

  int main(int argc, char* argv[]){


    int my_rank, nprocs;
    int i,j;
    int rem, num_local_row;
    int *matrix;
    int proc_up, proc_down;
    MPI_Status  status1, status2;

   /* Initialize the environment and store the size in "nprocs" and the rank in "my_rank" */
    ...

    /*  number of rows for each mpi task. How does it work? Try to write it down. */
    rem= N % num_procs;
    num_local_row = (N - rem)/nprocs;
    if(my_rank < rem) num_local_row = num_local_row+1;

    /* Allocation of the global matrix as an array of contiguous elements. Instead of "..." put one of the following:
    1. num_local_row  2. num_local_row+1  3. num_local_row+2 */
    matrix = (int*) malloc(((...)*N)*sizeof(int));

    /* initialization of the local matrix */
    for(i=0; i<num_local_row+2 ;i++){
      for(j=0; j<N; j++){
        matrix[i*N+j] = my_rank ;
      }
    }

    /* Information about the neighbour processes (imagine that you are sending up and down arrays of data, so the variables are named to reflect that) */
    proc_down = my_rank+1;
    proc_up = my_rank-1;
    if(proc_down==nprocs) proc_down=0;
    if(proc_up < 0) proc_up=nprocs-1;

    /* check printings */
    /* printf("rank %d, proc up %d, proc down %d\n",
    my_rank, proc_up, proc_down); */

    /* printf("my_rank %d, num_local_row %d\n", my_rank, num_local_row); */

    /* printf("my_rank %d,matrix[0] %d, matrix[N] %d,
    matrix[(N)*(num_local_row+2)] %d \n", my_rank, matrix[0],
    matrix[N], matrix[N*(num_local_row+2)]); */

    /* send receive of the ghost regions */
    /* First one: sends the uppest real row to the proc above it, and receives on the lowest ghost row from the proc below it */
    MPI_Sendrecv(...);

    /* Second one: sends the lowest real row to the prov below it, and receives on the uppest shost row from the proc above it */
    MPI_Sendrecv(...);


    /* check printings */
    printf("\n ");

    printf("my_rank, %d, riga arrivata da sopra: ", my_rank);
    for(i=0;i<N;i++){
      printf("%d \t", matrix[i]);
      }
    printf("\n ");

    printf("my_rank, %d, riga arrivata da sotto: ", my_rank);
    for(i=N*(num_local_row+1);i<N*(num_local_row+1)+N;i++){
      printf("%d \t", matrix[i]);
    }
    printf("\n ");

    free(matrix);
    /* Finalize MPI Environment */
    ...
    return 0;
  }
